### ISSUE TITLE
RDPC: Jenkins production provisions

### ISSUE THREAD

**Issue by akachru-github** [icgc-argo/roadmap#494]
- [ ] song
- [ ] workflow management
- [ ] workflow search
- [ ] workflow-relay
- [ ] workflow-ui
- [ ] workflow-docs
- [ ] workflow-gateway
- [ ] rdpc-status
- [ ] RDPC-Gateway
- [ ] RDPC Maestro
- [ ] intermediate-song
- [ ] score
- [ ] song-search
...list of services

We want to mirror platform task architecture in RDPC Prod Jenkins
- Create prod deploy task, parameterized build.
- Create task for current deployed versions, spits out pod image tags. 
- Investigate parameterized build details to Jenkins file. If possible, let's go ahead and do it. 

Testing
- Coordinate with Alex, document how to test in this ticket.

**Comment by yalturmes** [comment]
Progress so far...
* It is confirmed that Jenkinsfile can include parameters for parametrized build. The plan is to proceed configuring deploy job using parameters bundled in Jenkinsfile as opposed to having them configured via Jenkins UI. A draft block for implementing this has been prepared. Assuming further testing is successful, it will be necessary to retrofit this approach for defining build parameters into deploy job on ARGO Platform Jenkins instance.
* Jenkins credentials for accessing private repo, namely [rdpc-infra](https://github.com/icgc-argo/rdpc-infra/) have been configured on RDPC Prod Jenkins instance. A necessary pre-requisite.
* `Show Version` job has been configured. A follow up conversation with @andricDu has identified an additional item to look into. The `Show version` job on ARGO Platform Jenkins instance is configured as `Pipeline script` as apposed to `Pipeline script from SCM`. Using an SCM with Jenkinsfile is recommended to make the job setup reproducible/reusable in future.
* Conversation with @lepsalex helped identify plan for initial `deploy` job testing with minimal risk for RDPC PROD env. DEV/QA RDPC env will be used at the initial stage. After that the `deploy` job setup will be implemented and tested in prod environment. Issues with Kubernetes policies/RBAC/service account permissions expected once we progress to RDPC PROD env. Once ready for testing in PROD RDPC env - it will be necessary to communicate service disruptions in `workflow` channel.

**Comment by akachru-github** [comment]
Testing "manual deployment" Jenkins job being tested in DEV / QA now. Validating all deployment types before promoting the job to production Jenkins. 

Once in production: we will be performing a coordinated test of all manual deployment types in production.

**Comment by yalturmes** [comment]
* Jenkins file for the manual provision job has been updated to include DEV/QA and PROD environmnts. The Jenkinsfile script includes build parameters now. The Jenkinsfile script includes logic to workaround the inconsistent namespace naming. The same file will be used to configure jobs all RDPC Jenkins instances but only applicable environment options will be valid.
* Initial list of helm charts that need to be deployable manually have been updated to include `intermediate-song`, `score` and `song-search`.
* All Testing on DEV/QA env is done except for `song-search` chart as the chart version deployed on dev is ahead of the version available in helm repo. @lepsalex was made aware and will need to fix this to finish testing.
* RDPC PROD environment is used by the development team to run workflows between Jul 21-23rd. Manual deployment testing will have to be postponed until workflow jobs are completed.

**Comment by akachru-github** [comment]
We should consider doing these updates while the RDPC is busy as we will shooting for 100% utilization of the system. Schedule time to talk about this.

**Comment by yalturmes** [comment]
* The last remaining chart `song-search` have been tested in rdpc dev and qa environments. So this concludes ALL testing for manual-provision in RDPC DEV/QA.
* As per conversation with @andricDu (see details in earlier comments for this ticket), added a Jenkinsfile for show-versions job. To make the job re-usable in all environments, the Jenkinsfile includes parameter to select environment. This has been tested in RDPC DEV/QA/PROD.
* Created [pull request](https://github.com/icgc-argo/rdpc-infra/pull/108/files) with both Jenkinsfiles: show-versions and manual-provision.

Remaining items in the TODO list for this ticket: 
- [ ] Update branch references in both RDPC Jenkins instances once pull request is merged
- [ ] Re-configure show-versions job in ARGO PLATFORM PROD to use Jenkinsfile for setup consistency
- [ ] Discuss and test manual redeployment for all helm charts in RDPC PROD environment

**Comment by akachru-github** [comment]
Dusan's terraform pipeline work may have overridden this ticket. We will need to re-align with him on whether this ticket is still relevant or not. A lot has changed since we last worked on this in July 2020.

**Comment by akachru-github** [comment]
@yalturmes  to do an audit for where this ticket currently is. 
- Is this work actively being use din DEV / QA?
- Do we need to create followup tickets to update the work to account for changes to the RDPC?

**Comment by akachru-github** [comment]
After discussion with @yalturmes and @andricDu , this ticket is out of date and no longer applicable as the terraform changes Dusan is making will replace this. 
Closing.
